src/processor/operator/shortest_path_adj_list.cpp

```diff
@@ -0,0 +1,92 @@
+#include "processor/operator/shortest_path_adj_list.h"
+
+#include "iostream"
+
+using namespace std;
+
+namespace kuzu {
+namespace processor {
+
+bool ShortestPathAdjList::getNextTuplesInternal() {
+    while (true) {
+        computeShortestPath();
+        uint64_t currSrcIdx;
+        uint64_t currDestIdx;
+        do {
+            if (!children[0]->getNextTuple()) {
+                return false;
+            }
+            currSrcIdx = srcValueVector->state->getPositionOfCurrIdx();
+            currDestIdx = destValueVector->state->getPositionOfCurrIdx();
+        } while (srcValueVector->isNull(currSrcIdx) || destValueVector->isNull(currDestIdx) ||
+                 !addBFSLevel(srcValueVector->readNodeOffset(currSrcIdx), 1, UINT64_MAX));
+    }
+}
+
+void ShortestPathAdjList::computeShortestPath() {
```

Comment:
Let's simplify this function a bit by introducing more functions to decompose the computation logic.
```
// return false if no path can be found (within the lower and upper bound). otherwise return true.
bool ShortestPathAdjList::getShortestPath(nodeID_t srcNode, nodeID_t destNode) {
    // initialize the current frontier for layer 0.
    currentFrontier.addNode(UINT64_MAX, srcNode);
    // extend from the current frontier to the next one.
    while (currentLayer < upperBound) {
        destNode = currentLayer < lowerBound ? UINT64_MAX : destNode; // set destNodeID to invalid value if the layer is still less than the lower bound, so no matches with dest node can be found.
        if (extendToNextFrontier(destNode)) { return true; }
        // move vector in nextFrontier to currentFontier, and reset nextFrontier.
        resetFrontier();
        currentLayer++;
    }
    return false;
}
```

See the shared design doc for complete information.